Feature/zenko 203 md ingestion design (#453)
scality/backbeat: docs/oob-s3-ring.md

# Out-of-band (OOB) Updates from RING to Zenko

## Description

This design document describes the feature of ingesting metadata of objects put
in RING over S3 Connector into Zenko.

The development involves writing a service in Backbeat that allows

> ingesting existing objects as well as newly created objects from a remote

Reviewer: ```suggestion
the ingestion of existing objects as well as newly created objects from a remote
```

S3C/RING into Zenko.

## Purpose

> OOB updates from RING feature allows Zenko features on the configured RING

Reviewer: maybe it's just me nitpicking, but I feel like it should be stressed that the OOB feature will allow for Zenko features to work on objects that have been put or will be `PUT` outside of Zenko rather than the RING bucket.

> bucketsuch as CRR, Lifecycle Expiration, Lifecycle Transition, and Metadata

Reviewer 2: bucket such

> Search. This enables usecases such as

Reviewer 2: use cases

- Cloud D/R i.e. backing up RING to Zenko supported cloud locations
- Policy based transition of objects to cloud locations
- Policy based expiring of existing non-current (archive) versions
- Searching user-defined and system metadata on objects

## Design

The feature allows ingesting objects by the configured RING bucket into the
mapped Zenko bucket.

> ### Proposed workflow

Reviewer 3: I'm not sure about the arrow pointing from Cloudserver to Backbeat

Reviewer 4: You are right.

|          |      | |             | +------------------------------+
+----------+   +-----+ Zookeeper  | |  Using the list of object keys,
               |  | |             | |  get the JSON of each object from Metadata
               |  | +-------------+ |
               |  |                 |
               |  |    Backbeat     |
               |  |                 |
               |  +-----------------+
               |
               |     On startup, check Zookeeper to
               +---> see if a 'sequence id' or `Next Continuation Token` already
                     exists.

```

* Create a new S3C location in Orbit by adding a versioning-enabled S3C bucket
  and `bucketMatch`  option set to true. This ensures that objects appear
  natively without any prefixes.
* Create a bucket in Zenko and select the S3C location created in the above step
  as location constraint that is tagged as `Mirror mode`. This will setup the
  ingestion workflow on the Zenko bucket, allowing 1-1 mapping between the
  bucket defined through Zenko and the RING bucket.
* Cloudserver tags the bucket with the attribute
  `ingestion: { status: 'enabled'}`

> * Backbeat will have a service that monitors the MongoDB oplog for the buckets

Reviewer: ```suggestion
* Backbeat will have a service that monitors the MongoDB oplog for buckets
```

>   that have ingestion enabled using the `ingestion` attribute and setups up

Reviewer 2: sets up

  the ingestion workflow.
* The ingestion process is composed of two phases:
    - **Initial ingestion of existing objects**
    - **Out of band updates for new objects**

#### Initial ingestion of existing objects

* Record the latest sequence ID from the Metadata Server (send a query using the
  `metadataWrapper`) that is part of the S3 connector - this will serve as a
  marker that will be stored as a `log offset` to bookmark the point in journal
  to resume later, after the initial ingestion is completed.
* Send List Object Versions API request to S3C to receive a list of

>   object versions in the bucketwith details like  `key` and `versionId`.

Reviewer 2: bucket with

> * Retrieve object JSON from S3C Metadata with the `key` and `versionId`.

Reviewer 3: ~I'm not sure about the necessity to make another roundtrip to bucketd as the metadata of individual objects is already present in the listing. (Or maybe I didn't understand right and we're not talking to bucketd here?)~

Edit: Yes I was wrong. buckets are listed through the S3 protocol. Why don't we list using directly the Metadata service as we're willing to access it anyway?

Reviewer 4: One of the main reasons is that we need a continuation token of some sort if the listing is aborted in the middle to resume from where it was. S3 api's listing makes it very easy to manage this, and the same logic can be re-used for other S3-compatible clouds.

* `Next Continuation Token` from listings is used to keep track of listing, to
  recover during unexpected pod/node failures

**Note: Requests to the remote S3C Metadata are performed over S3C's Cloudserver
which proxies the request to Metadata. The request is secure and signed by

> AWS V4 authentication scheme.**

Reviewer 3: Oh! I understand now, Zenko and S3C are probably not in the same zone! Right?

Reviewer 4: They may or may not be but we assume that they are completely different zones.

#### Out of band updates for new objects

> After initial ingestion is completed, Ingestion service will use it's producer

Reviewer: ```suggestion
After the initial ingestion has completed, Ingestion service will use it's producer
```

to read from the remote S3C's Metadata and queue the journal entries to Kafka.
Ingestion consumer picks up the entry from Kafka and processes the following
fields for object metadata:

* `owner`
* `ownerDisplayName`
* `owner-id`
* `location` if defined

> * `dataStoreName` should be set to `scality`

Reviewer 2: dataStoreName should be set to the Zenko location storage name

operations by the Orbit user.

A Zenko deployment will have 1 `ingestion-populator` pod and multiple
`ingestion-processor` pods. The number of ingestion processor pods depends on
the `nodeCount` value specified during deployment (default is 3). The
`ingestion-populator` pod can support multiple ingestion workflows. The
logReader for each ingestion populator will be setup to source logs for the
specified ingestion bucket.

## Dependencies

* RING with S3 Connector on version 7.4.2 or newer
* Backbeat
* MongoDB
* Kafka
* Zookeeper

## Bucket/Object-Key Parity

The ingestion source is a bucket on RING registered as a storage location with
`bucketMatch` enabled. This ensures an exact match of object names without any
prefixes (unlike the prefix schema used in Zenko locations that host multiple
Zenko 'buckets').

## Avoiding Retro Propagation

The S3C location in `mirror mode` is added as a location constraint to the
Zenko bucket. This allows putting objects in the Zenko bucket, with object's
metadata stored in Zenko, and object's metadata & data stored in S3C.
To avoid re-ingesting these objects back into Zenko, objects will be assigned

> `scal-source: zenko-<instance id>` in thier in their user-metadata. Any object

Reviewer 2: ~~thier~~ their

> `scal-source: zenko-<instance id>` in their in their user-metadata. Any object

Reviewer 3: 'in their' twice here.

As a note: Is Avoiding Retro Propagation, the prevention against considering a PUT through Zenko as an out of band update? right?

Reviewer 4: > As a note: Is Avoiding Retro Propagation, the prevention against considering a PUT through Zenko as an out of band update? right?

Yup, that's correct

> verions that have this attribute in their user-metadata will be ignored during

Reviewer 2: ~~verions~~ versions
